modules/ray_tracing/src/userobjects/RepeatableRayStudy.C

```diff
@@ -0,0 +1,161 @@
+                              ? &getParam<std::vector<Real>>("initial_ray_aux_data")
+                              : nullptr)
+{
+  for (std::size_t i = 0; i < _ids.size(); ++i)
+    if (_ids[i] == Ray::INVALID_RAY_ID)
+      paramError("names",
+                 "Ray name '",
+                 getParam<std::string>("names")[i],
+                 "' could not be registered. It may have been provided multiple times.");
+  if (_end_points && _directions)
+    paramError("directions", "Can only use 'directions' or 'end_points', but not both");
+  if (_start_points.size() != _ids.size())
+    paramError("start_points", "Not the same size as names");
+  if (_directions && _ids.size() != _directions->size())
+    paramError("directions", "Not the same size as names");
+  if (_max_distances)
+  {
+    if (!_directions)
+      paramError("max_distances",
+                 "Can only be used when trajectories are set with the 'directions' parameter");
+    if (_max_distances->size() != _start_points.size())
+      paramError("max_distances", "Must be the same size as 'start_points'");
+    for (const auto val : *_max_distances)
+      if (val <= 0)
+        paramError("max_distances", "Values must be positive");
+  }
+  if (_end_points && _ids.size() != _end_points->size())
+    paramError("end_points", "Not the same size as names");
+  if (!_ray_data_indices.empty() && _initial_ray_data &&
+      _initial_ray_data->size() != _ray_data_indices.size())
+    paramError("initial_ray_data", "Not the same size as ray_data_names");
+  if (_initial_ray_data && _ray_data_indices.empty())
+    paramError("initial_ray_data", "Can only be used if ray_data_names is set");
+  if (!_ray_aux_data_indices.empty() && _initial_ray_aux_data &&
+      _initial_ray_aux_data->size() != _ray_aux_data_indices.size())
+    paramError("initial_ray_aux_data", "Not the same size as ray_aux_data_names");
+  if (_initial_ray_aux_data && _ray_aux_data_indices.empty())
+    paramError("initial_ray_aux_data", "Can only be used if ray_aux_data_names is set");
+}
+
+void
+RepeatableRayStudy::defineRays()
+{
+  for (std::size_t i = 0; i < _ids.size(); ++i)
+  {
+    std::shared_ptr<Ray> ray = acquireRay(/* tid = */ 0);
+
+    if (_end_points) // user set end point
+      ray->setStartEnd(_start_points[i], (*_end_points)[i]);
+    else // user set direction
+      ray->setStartDirection(_start_points[i], (*_directions)[i]);
+
+    // We already generated the IDs
+    ray->setID(_ids[i]);
+
+    // Set the data
+    ray->data().resize(rayDataSize(), 0);
+    if (_initial_ray_data)
+      for (std::size_t i = 0; i < _ray_data_indices.size(); ++i)
+        ray->data(_ray_data_indices[i]) = (*_initial_ray_data)[i];
+
+    // Set the aux data
+    ray->auxData().resize(rayAuxDataSize(), 0);
+    if (_initial_ray_aux_data)
+      for (std::size_t i = 0; i < _ray_aux_data_indices.size(); ++i)
+        ray->auxData(_ray_aux_data_indices[i]) = (*_initial_ray_aux_data)[i];
```

Comment:
I think all this could be handled by the registerRayData/AuxData functions.  They could take optional initial data and initial aux data args, resize everything, and set the initial stuff.  Then this sort of "user" code won't have to muck around with keeping track of IDs, consistent resizing, etc.  Or you could pass it in as arguments to the acquireRay function.  Basically I think we could make this entire function look like:

```c++
    auto & ray = acquireRay(_tid, initial_data, initial_aux_data);
    if (_end_points) // user set end point
      ray->setStartEnd(_start_points[i], (*_end_points)[i]);
    else // user set direction
      ray->setStartDirection(_start_points[i], (*_directions)[i]);
    if (_max_distances)
      ray->setMaxDistance((*_max_distances)[i]);
    _rays.emplace_back(std::move(ray));
```

Where there are way less ways to write it wrong.  And sizing-consistency, etc. is enforced more strongly because it is all controlled from one location.